plat/common/include/arm/arm64/paging.h

```diff
@@ -0,0 +1,306 @@
+
+	if (!(pte & ATTR_AP(ATTR_AP_RO)))
+		attr |= PAGE_ATTR_PROT_WRITE;
+
+	if (!(pte & ATTR_PXN))
+		attr |= PAGE_ATTR_PROT_EXEC;
+
+	return attr;
+}
+
+/*
+ * Page tables
+ */
+static inline __vaddr_t
+pgarch_pt_pte_to_vaddr(struct uk_pagetable *pt __unused, __pte_t pte,
+		       unsigned int level __maybe_unused)
+{
+	(void)pt;
+
+	__vaddr_t ret = arm64_directmap_paddr_to_vaddr(PT_Lx_PTE_PADDR(pte, level));
+
+	return ret;
+}
+
+/* Create a table descriptor */
+static inline __pte_t
+pgarch_pt_pte_create(struct uk_pagetable *pt __unused, __paddr_t pt_paddr,
+		     unsigned int level __unused, __pte_t template,
+		     unsigned int template_level __unused)
+{
+	(void)template;
+
+	__pte_t pt_pte;
+
+	UK_ASSERT(PAGE_ALIGNED(pt_paddr));
+
+	pt_pte = pt_paddr & PTE_Lx_TABLE_PADDR_MASK;
+
+	pt_pte |= Ln_TABLE;
+
+	return pt_pte;
+}
+
+static inline __vaddr_t
+pgarch_pt_map(struct uk_pagetable *pt __unused, __paddr_t pt_paddr,
+	      unsigned int level __unused)
+{
+	return arm64_directmap_paddr_to_vaddr(pt_paddr);
+}
+
+static inline __paddr_t
+pgarch_pt_unmap(struct uk_pagetable *pt __unused, __vaddr_t pt_vaddr,
+		unsigned int level __unused)
+{
+	return arm64_directmap_vaddr_to_paddr(pt_vaddr);
+}
+static inline int pgarch_init(void)
+{
+	/* Sanity checks to make sure that the PE supports the minimum
+	 * requirements and the platform has set up a sane environment.
+	 * This is expected to be useful during bring-up, so keep it
+	 * conditional to UKDEBUG, to improve boot performance.
+	 */
+#ifdef CONFIG_LIBUKDEBUG
+	unsigned int ia_size;
+	unsigned int oa_size;
+	unsigned int tgran4;
+
+	__u64 reg = SYSREG_READ64(TCR_EL1);
+
+	/* Check 48-bit addressing mode configured */
+	if (reg & TCR_EL1_DS_BIT)
+		UK_CRASH("48-bit addressing in not enabled\n");
+
+	/* Check 48-bit IA configured. */
+	ia_size = 64 - ((reg & TCR_T0SZ_MASK) >> TCR_T0SZ_SHIFT);
+
+	if (ia_size != ARM64_PADDR_BITS)
+		UK_CRASH("Invalid paddr width: %d bits\n", ia_size);
+
+	/* Check 48-bit OA configured */
+	oa_size = tcr_ips_bits[(reg >> TCR_IPS_SHIFT) & TCR_IPS_MASK];
+
+	if (oa_size != ARM64_VADDR_BITS)
+		UK_CRASH("Invalid vaddr width: %d bits", oa_size);
+
+	/* Check 4K granule size configured */
+	if ((reg & (TCR_EL1_TG0_MASK << TCR_EL1_TG0_SHIFT)) != TCR_EL1_TG0_4K)
+		UK_CRASH("4KiB granule size is not enabled\n");
+
+	/* Check TTBR0 walks are enabled */
+	if (reg & TCR_EL1_EPD0_BIT)
+		UK_CRASH("TTBR0_EL1 table walks are not enabled\n");
+
+	/* Check TTBR1 walks are disabled */
+	if (!(reg & TCR_EL1_EPD1_BIT))
+		UK_CRASH("TTBR1_EL1 table walks are not disabled\n");
+
+	/* Check 4K granule size supported */
+	reg = SYSREG_READ64(ID_AA64MMFR0_EL1);
+
+	tgran4 = (reg >> ID_AA64MMFR0_EL1_TGRAN4_SHIFT) &
+		  ID_AA64MMFR0_EL1_TGRAN4_MASK;
+
+	if (tgran4 != ID_AA64MMFR0_EL1_TGRAN4_SUPPORTED)
+		UK_CRASH("4KiB granule not supported\n");
+#endif /* CONFIG_LIBUKDEBUG */
+	return 0;
+}
+
+static inline int
+pgarch_pt_add_mem(struct uk_pagetable *pt, __paddr_t start, __sz len)
```

Comment:
Maybe in this case, but doesn't mean it will apply to any processor or any implementation (eg not every implementation has to use direct mapped memory), so maybe we should skip the temptation to overfit.